CPU level UDP packet counter

ykyohei opened the pull request:
Add CPU level UDP packet counter to the HWP encoder.

To identify whether we are dropping the hwp encoder packet at PRU->CPU level or at network level (UDP packet loss), we add the CPU level packet counter. The `counter_index` is indexed at the PRU level.

This has not been tested yet. We need to use an updated Encoder Agent together.

ykyohei committed:
Add cpu level packet counter

ykyohei committed:
fill packet counter

ykyohei committed:
fix typo

ykyohei committed:
Revert "fix typo"

This reverts commit 67894a03902d115d4f30967c7a11c694502f4a0a.

jlashner committed:
bugfix in packet counter

jlashner committed:
add buffer reset counter and num_samps to packet

ykyohei commented on hwp_encoder/Beaglebone_Encoder_DAQ.c:
```diff
@@ -283,12 +292,17 @@ int main(int argc, char **argv) {
         }
         // Send encoder data if the buffer is full
         if(encd_ind == ENCODER_PACKETS_TO_SEND) {
+            for (int i = 0; i < ENCODER_PACKETS_TO_SEND; i++){
+                encoder_to_send[i].packet_count = encd_pkt_cnt++;
+            }
             sendto(sockfd, (struct EncoderInfo *) encoder_to_send, sizeof(encoder_to_send), 
                    MSG_CONFIRM, (const struct sockaddr *) &servaddr, sizeof(servaddr));
             encd_ind = 0;
         }
 	// Send IRIG data if the buffer is full
         if(irig_ind == IRIG_PACKETS_TO_SEND) {
+            irig_pkt_cnt += 1;
+            irig_to_send[irig_ind].packet_count = irig_pkt_cnt;
```
I wonder why we increment packet_count differently in encoder and irig?
Something is wrong with encoder_packet, because i is always 0?

jlashner replied:
Good point, I think this is wrong for the irig packet counter, but I have not been checking that in my tests.

ykyohei reviewed (commented):
Thanks for the updates, I have one comment.

ykyohei commented:
Another comment/question.
It might be fine to have more information, but I'm bit confused by the difference between `buffer_reset_counter` and `packet_counter`. They seem almost redundant to me.
Your `buffer_reset_counter` seems more fundamental, so we can remove the `packet_counter`?

Then we should have `buffer_reset_counter` also for the IRIG packet.

jlashner commented:
Ya so the main difference between the counters is that the `buffer_reset_counter` counts when the PRU has finished writing a packet and starts on the next one, while `packet_counter` increments when the CPU detects a new packet and sends it over the network.

These are related but we still want to track both of them. The failure mode that I am concerned about is when the CPU does not register the packet from the PRU in time to write it to the network before the PRU overwrites it. This will show up as a dropped index in the `buffer_reset_counter`, while there is no corresponding dropped index in the `packet_counter` (meaning the packet was never written to the network).  Like in this plot below:
![image](https://github.com/simonsobs/sobonelib/assets/4718487/b36a1c39-2926-40f8-952f-24b8d8f23ea5)


I don't think I know enough about how IRIG is handled to know if we need the same kind of book-keeping. Do you know how often the IRIG buffer is swapped / if we're concerned about that dropping?

ykyohei commented:
Ah okay thanks.
If I understand correctly, `reset_buffer_counter` is redundant as the 120 jump in `counter`, but I agree that it is helpful to have `reset_buffer_counter`.

I was also less careful about IRIG data. Unfortunately, we have jump in IRIG time as well, and I think the rate is higher than in lab. This might be because of the site timing system. Also, this is less concerning because the interpolation of the dropped IRIG time is much simpler, but I think it is definitely helpful to monitor.
![Figure 6 (1)](https://github.com/simonsobs/sobonelib/assets/38639108/6a3630b6-be36-4a8e-913c-3bd81a6bb64a)

The rate is not negligible. `WARNING of deleted wrong irig_time` corresponds to this issue.

```
16774.603: Calculating Angles for obs_1710841702_satp1_1111111 (INFO)
WARNING: sotodlib.core.context: Key "metadata" not present in context.
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [1199 1842 2410]
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 6 dropped packets are found, performing fill process
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [1199]
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 2 dropped packets are found, performing fill process
17046.032: Calculating Angles for obs_1710848051_satp1_1111111 (INFO)
WARNING: sotodlib.core.context: Key "metadata" not present in context.
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 1 dropped packets are found, performing fill process
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 1 dropped packets are found, performing fill process
17159.491: Calculating Angles for obs_1710865890_satp1_1101111 (INFO)
WARNING: sotodlib.core.context: Key "metadata" not present in context.
WARNING: sotodlib.hwp.g3thwp: No HWP data in the specified timestamps.
WARNING: sotodlib.hwp.g3thwp: No HWP data in the specified timestamps.
WARNING: sotodlib.hwp.g3thwp: Offcentering calculation is only available when two encoders are operating. Skipped.
17635.181: Calculating Angles for obs_1710882494_satp1_1101111 (INFO)
WARNING: sotodlib.core.context: Key "metadata" not present in context.
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [ 643 1239 1826 1831 4022 4027]
WARNING: sotodlib.hwp.g3thwp: cannot find reference points, # of data is less than # of slit
WARNING: sotodlib.hwp.g3thwp: analyzed encoder data is None
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [ 643 1239 1833 3399]
WARNING: sotodlib.hwp.g3thwp: cannot find reference points, # of data is less than # of slit
WARNING: sotodlib.hwp.g3thwp: analyzed encoder data is None
WARNING: sotodlib.hwp.g3thwp: Offcentering calculation is only available when two encoders are operating. Skipped.
17884.236: Calculating Angles for obs_1710897250_satp1_1111111 (INFO)
WARNING: sotodlib.core.context: Key "metadata" not present in context.
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [  28 1226]
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 1 dropped packets are found, performing fill process
WARNING: sotodlib.hwp.g3thwp: deleted wrong irig_time, indices: [624]
WARNING: sotodlib.hwp.g3thwp: Packet drop exists
WARNING: sotodlib.hwp.g3thwp: 1 dropped packets are found, performing fill process
```

jlashner commented:
Yes... kind of. It's redundant with 120-edge frame drops as long as long as samples can only be dropped in complete packets. I think there are also some race conditions in the code which would mean these counters aren't completely redundant, but I believe those are much more unlikely to be triggered.

ykyohei committed:
fix irig packet_count

jlashner committed:
whitespace fixes

jlashner committed:
add version field and change header for irig packets

bbixler500 committed:
Fixed indexing error

bbixler500 commented:
I'm fine merging this code on the beaglebone side. However, I believe the agent code branch needs to be updated as well before we do that. Currently, when I look at the `hwp_encoder_packet_count` branch on `socs` the code is not set up currently to handle the `version`, `num_samples`, and `buffer_reset_counter` fields.

jlashner commented:
There is currently a PR here that adds that support: https://github.com/simonsobs/socs/pull/647

I think there are possibly some issues remaining, but I was having some trouble emulating irig data packets locally which is why I ended up putting it on hold. I think it would still be good to get this working though.
